shell/store/notifications.ts

```diff
@@ -0,0 +1,272 @@
+import { randomStr } from '@shell/utils/string';
+
+const LOCAL_STORAGE_KEY = '#notifications';
+
+// Expire in seconds (14 days)
+const EXPIRY = 14 * 24 * 60 * 60;
+
+// Maximum number of notifications that will be kept
+const MAX_NOTIFICATIONS = 50;
+
+/**
+ * Store for the UI Notification Centre
+ */
+
+export type NotificationAction = {
+  label: string; // Button label for the action
+  target?: string; // HREF target when the button is clicked
+};
+
+export type NotificationPreference = {
+  key: string; // User preference key to use when setting the preference when the notification is marked as read
+  value: string; // User preference value to use when setting the preference when the notification is marked as read
+  unsetValue?: string; // User preference value to use when setting the preference when the notification is marked as unread - defaults to empty string
+};
+
+export enum NotificationLevel {
+  Announcement = 0, // eslint-disable-line no-unused-vars
+  Task, // eslint-disable-line no-unused-vars
+  Info, // eslint-disable-line no-unused-vars
+  Success, // eslint-disable-line no-unused-vars
+  Warning, // eslint-disable-line no-unused-vars
+  Error, // eslint-disable-line no-unused-vars
+}
+
+/**
+ * Type for Notification that is sent
+ */
+export type Notification = {
+  // Unique ID for the notification
+  id: string;
+  title: string;
+  message?: string;
+  level: NotificationLevel;
+  progress?: number;
+  primaryAction?: NotificationAction;
+  secondaryAction?: NotificationAction;
+  preference?: NotificationPreference;
+}
+
+/**
+ * Type for notification that is stored
+ *
+ * Includes extra fields managed by the notification center
+ */
+export type StoredNotification = {
+  created: Date;
+  read: Boolean;
+} & Notification;
+
+// Notifications store state
+interface NotificationsStore {
+  notifications: StoredNotification[],
+}
+
+function persist(state: NotificationsStore) {
+  window.localStorage.setItem(LOCAL_STORAGE_KEY, JSON.stringify(state.notifications));
+}
+
+export const state = function(a: any): NotificationsStore {
+  let notifications = [];
+
+  try {
+    notifications = JSON.parse(window.localStorage.getItem(LOCAL_STORAGE_KEY) || '[]');
+  } catch (e) {
+    console.error('Unable to read notifications from local storage', e); // eslint-disable-line no-console
+  }
+
+  // Expire old notifications
+  const now = new Date();
+
+  notifications = notifications.filter((n: StoredNotification) => {
+    // Try ... catch in case the date parsing fails
+    try {
+      const created = new Date(n.created);
+      const diff = (now.getTime() - created.getTime()) / 1000; // Diff in seconds
+
+      return diff < EXPIRY;
+    } catch (e) {}
+
+    return true;
+  });
+
+  return { notifications };
+};
```

Comment:
Similar to a data prop in a vue component, a store's state should avoid complex logic to focus on declaring reactive state. Would this logic initialization logic better fit in the `init()` action?

Reply:
Good idea - done